geomesa-kafka/geomesa-kafka-09-datastore/src/main/scala/org/locationtech/geomesa/kafka09/LiveKafkaConsumerFeatureSource.scala

```diff
@@ -0,0 +1,276 @@
+
+  def buildBounds(feature: SimpleFeature): ReferencedEnvelope = {
+    try {
+      val geom = feature.getDefaultGeometry.asInstanceOf[Point]
+      val lon = geom.getX
+      val lat = geom.getY
+
+      ReferencedEnvelope.create(new Envelope(lon, lon, lat, lat), CRS_EPSG_4326)
+    } catch {
+      case t: Throwable =>
+        KafkaConsumerFeatureSource.wholeWorldBounds
+    }
+  }
+
+  def changed(src: SimpleFeatureSource, feature: SimpleFeature): FeatureEvent =
+    new KafkaFeatureEvent(this,
+      Type.CHANGED,
+      KafkaFeatureEvent.buildBounds(feature),
+      feature)
+
+  def removed(src: SimpleFeatureSource, feature: SimpleFeature): FeatureEvent =
+    new FeatureEvent(this,
+      Type.REMOVED,
+      KafkaFeatureEvent.buildBounds(feature),
+      KafkaFeatureEvent.buildId(feature.getID))
+
+  def cleared(src: SimpleFeatureSource): FeatureEvent =
+    new FeatureEvent(this,
+      Type.REMOVED,
+      KafkaConsumerFeatureSource.wholeWorldBounds,
+      Filter.INCLUDE)
+}
+
+/** @param sft the [[SimpleFeatureType]]
+  * @param expirationPeriod the number of milliseconds after write to expire a feature or ``None`` to not
+  *                         expire
+  * @param ticker used to determine elapsed time for expiring entries
+  */
+class LiveFeatureCache(override val sft: SimpleFeatureType,
+                       expirationPeriod: Option[Long])(implicit ticker: Ticker)
+  extends KafkaConsumerFeatureCache with LazyLogging {
+
+  def cleanUp(): Unit = cache.cleanUp()
+
+  var spatialIndex: SpatialIndex[SimpleFeature] = newSpatialIndex()
+
+  val cache: Cache[String, FeatureHolder] = {
+    val cb = CacheBuilder.newBuilder().ticker(ticker)
+    expirationPeriod.foreach { ep =>
+      cb.expireAfterWrite(ep, TimeUnit.MILLISECONDS)
+        .removalListener(new RemovalListener[String, FeatureHolder] {
+          def onRemoval(removal: RemovalNotification[String, FeatureHolder]) = {
+            if (removal.getCause == RemovalCause.EXPIRED) {
+              logger.debug(s"Removing feature ${removal.getKey} due to expiration after ${ep}ms")
+              spatialIndex.remove(removal.getValue.env, removal.getValue.sf)
+            }
+          }
+        })
+    }
+    cb.build()
+  }
+
+  override val features: mutable.Map[String, FeatureHolder] = cache.asMap().asScala
+
+  def createOrUpdateFeature(update: CreateOrUpdate): Unit = {
+    val sf = update.feature
+    val id = sf.getID
+    val old = cache.getIfPresent(id)
+    if (old != null) {
+      spatialIndex.remove(old.env, old.sf)
+    }
+    val env = sf.geometry.getEnvelopeInternal
+    spatialIndex.insert(env, sf)
+    cache.put(id, FeatureHolder(sf, env))
+  }
+
+  def removeFeature(toDelete: Delete): Unit = {
+    val id = toDelete.id
+    val old = cache.getIfPresent(id)
+    if (old != null) {
+      spatialIndex.remove(old.env, old.sf)
+      cache.invalidate(id)
+    }
+  }
+
+  def clear(): Unit = {
+    cache.invalidateAll()
+    spatialIndex = newSpatialIndex()
```

Comment:
I'd probalby change the method name to createSpatialIndex() since it looks like it was supossed to be new [SPACE] SpatialIndex()